server/core/dcb.cc

```diff
@@ -888,6 +895,13 @@ dcb_write(DCB *dcb, GWBUF *queue)
     dcb->stats.n_buffered++;
     dcb_drain_writeq(dcb);
 
+    if (DCB_ABOVE_HIGH_WATER(dcb) && !dcb->high_warter_has_reached)
```

Comment:
I think with something like this the router API changes could be removed:

```cpp
bool add_func(DCB *dcb, void *data)
{
    MXS_SESSION* session = (MXS_SESSION*)data;

    if (dcb->session == session && dcb->dcb_role == DCB_ROLE_BACKEND_HANDLER)
    {
        poll_add_dcb(dcb);
    }

    return true;
}

bool remove_func(DCB *dcb, void *data)
{
    MXS_SESSION* session = (MXS_SESSION*)data;

    if (dcb->session == session && dcb->dcb_role == DCB_ROLE_BACKEND_HANDLER)
    {
        poll_remove_dcb(dcb);
    }

    return true;
}

/* Code in dcb.cc:898 */

    if (DCB_ABOVE_HIGH_WATER(dcb) && !dcb->high_water_has_reached)
    {
        dcb_foreach(remove_func, dcb->session)
        dcb->high_warter_has_reached = true;
        dcb->stats.n_high_water++;
    }

/* ... snip ... */

    if (dcb->high_warter_has_reached && DCB_BELOW_LOW_WATER(dcb))
    {
        dcb_foreach(add_func, dcb->session);
        dcb->high_warter_has_reached = false;
        dcb->stats.n_low_water++;
    }
```
Further improvements to efficiency can be done by linking the DCBs into the MXS_SESSION object. This can be done as a part of general refactoring work for the core.